rma/models/stock_move.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2020 Tecnativa - Ernesto Tejeda
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    # RMAs that were created from the delivery move
+    rma_ids = fields.One2many(
+        comodel_name="rma",
+        inverse_name="move_id",
+        string="RMAs",
+        copy=False,
+    )
+    # RMAs linked to the incoming movement from client
+    rma_receiver_ids = fields.One2many(
+        comodel_name="rma",
+        inverse_name="reception_move_id",
+        string="RMA receivers",
+        copy=False,
+    )
+    # RMA that create the delivery movement to the customer
+    rma_id = fields.Many2one(
+        comodel_name="rma",
+        string="RMA return",
+        copy=False,
+    )
+
+    def unlink(self):
+        # A stock user could have no RMA permissions, so the ids wouldn't
+        # be accessible due to record rules.
+        rma_receiver = self.sudo().mapped("rma_receiver_ids")
+        rma = self.sudo().mapped("rma_id")
+        res = super().unlink()
+        rma_receiver.write({"state": "draft"})
+        rma.update_received_state()
+        rma.update_replaced_state()
+        return res
+
+    def _action_cancel(self):
+        res = super()._action_cancel()
+        # A stock user could have no RMA permissions, so the ids wouldn't
+        # be accessible due to record rules.
+        cancelled_moves = self.filtered(lambda r: r.state == "cancel").sudo()
+        cancelled_moves.mapped("rma_receiver_ids").write({"state": "draft"})
+        cancelled_moves.mapped("rma_id").update_received_state()
+        cancelled_moves.mapped("rma_id").update_replaced_state()
+        return res
+
+    def _action_done(self, cancel_backorder=False):
+        """Avoids to validate stock.move with less quantity than the
+        quantity in the linked receiver RMA. It also set the appropriated
+        linked RMA to 'received' or 'delivered'.
+        """
+        for move in self.filtered(lambda r: r.state not in ("done", "cancel")):
+            rma_receiver = move.sudo().rma_receiver_ids
+            if rma_receiver and move.quantity_done != rma_receiver.product_uom_qty:
+                raise ValidationError(
+                    _(
+                        "The quantity done for the product"
+                        " '%(move.product_id.name)s' must be equal to its"
+                        " initial demand because the stock move is linked to"
+                        " an RMA (%(move.rma_receiver_ids.name)s)."
+                    )
+                )
```

Comment:
```suggestion
                raise ValidationError(
                    _(
                        "The quantity done for the product '%(id)s' must "
                        "be equal to its initial demand because the "
                        "stock move is linked to an RMA (%(name)s)."
                    )
                    % ({"id": move.product_id.name, "name": move.rma_receiver_ids.name})
                )
```